GraphResponse.md

```diff
@@ -7,6 +7,206 @@ Provide a decorator for the native response message to enable attaching Graph sp
 ## Requirements
 
 - Maintain reference to corresponding GraphRequest object
+- Provide access to HTTP response headers, status codes, and the raw response body.
 - Enable transformation from GraphResponse object to native response object.  This may be lossy.
 - Enable transformation from native response object to GraphResponse object
-- Provide access to response content as domain object.
\ No newline at end of file
+- Provide access to response content as domain object.
+- Provide a method to lazy deserialize the response body into our generated models or another object type based on whether a custom deserializer is provided.
+- Possilbly provide support providing a custom deserializer to override the default deserializer.
+
+## Performance Considerations
+
+Lazy deserialization should improve performance for scenarios where customers do not need to immediately access the returned objects.
+Custom deserializer support can enable customers to craft a deserializer optimized for a specific response.
+
+## Background
+
+The generated .Net client library either returns: 
+ 1. the response content as an object or collection object as the result of a successful call to Microsoft Graph, or 
+ 2. a ServiceException in the case the call fails.
+
+We consider this a design flaw as the client library swallows useful information such as the response headers and the http status code. Mitigation such as stuffing the information in AdditionalData is insufficient for all scenarios. 
+For example, there is no AdditionalData on Stream objects, and Delete operations return void.
+
+Related to:
+
+- https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/457
+- https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/361
+
+
+## Current Implementations
+
+### Javascript
+
+The Javascript SDK exposes the native [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) object when selecting the RAW ResponseType option.
+
+### .Net
+
+> Currently not supported
+
+### Java
+
+> Currently not supported. Should be able to use the same pattern as .Net.
+
+### PHP
+
+The PHP SDK returns a Graph model (or collection) or a GraphResponse object, depending on whether a returnType is specified. This means that the returnType information is not supplied to the GraphResponse. The response body is set as a JSON array on GraphResponse. The only work I'd suggest here is that we add an option to pass the returnType to the GraphResponse and add a lazy deserialize method so it is easy to get the Graph models.
+
+## The GraphResponse Object
+
+The GraphResponse object can wrap the native HttpResponseMessage object so that customers can use the familiar HttpResponseMessage members. 
+GraphResponse will be part of Microsoft.Graph.Core and it will expose the headers,Content and status code while also keeping reference to the Request object as below.
+
+```csharp
+
+    /// <summary>
+    /// The GraphResponse Object
+    /// </summary>
+    public class GraphResponse 
```

Comment:
We can make this generic to provide a type parameter deserialization hint. We already know what the response body content is when generating the code files so we can generate the type parameter at that time. I don't think there are any type constraints to be set.

Reply:
I have added a generic version of GraphResponse that derives from GraphResponse. In the case of delete operations we can return the non generic versions.